python/pyspark/sql/functions.py

```diff
@@ -2733,6 +2733,33 @@ def udf(f=None, returnType=StringType()):
+    #
+    # +-----------------------------+--------------+----------+------+-------+------+----------+--------------------+-----------------------------+----------+----------------------+---------+--------------------+--------------+----------+--------------+-------------+-------------+  # noqa
+    # |SQL Type \ Python Value(Type)|None(NoneType)|True(bool)|1(int)|1(long)|a(str)|a(unicode)|    1970-01-01(date)|1970-01-01 00:00:00(datetime)|1.0(float)|array('i', [1])(array)|[1](list)|         (1,)(tuple)|ABC(bytearray)|1(Decimal)|{'a': 1}(dict)|Row(a=1)(Row)|Row(a=1)(Row)|  # noqa
+    # +-----------------------------+--------------+----------+------+-------+------+----------+--------------------+-----------------------------+----------+----------------------+---------+--------------------+--------------+----------+--------------+-------------+-------------+  # noqa
+    # |                         null|          None|      None|  None|   None|  None|      None|                None|                         None|      None|                  None|     None|                None|          None|      None|          None|            X|            X|  # noqa
+    # |                      boolean|          None|      True|  None|   None|  None|      None|                None|                         None|      None|                  None|     None|                None|          None|      None|          None|            X|            X|  # noqa
+    # |                      tinyint|          None|      None|     1|      1|  None|      None|                None|                         None|      None|                  None|     None|                None|          None|      None|          None|            X|            X|  # noqa
+    # |                     smallint|          None|      None|     1|      1|  None|      None|                None|                         None|      None|                  None|     None|                None|          None|      None|          None|            X|            X|  # noqa
+    # |                          int|          None|      None|     1|      1|  None|      None|                None|                         None|      None|                  None|     None|                None|          None|      None|          None|            X|            X|  # noqa
+    # |                       bigint|          None|      None|     1|      1|  None|      None|                None|                         None|      None|                  None|     None|                None|          None|      None|          None|            X|            X|  # noqa
+    # |                       string|          None|      true|     1|      1|     a|         a|java.util.Gregori...|         java.util.Gregori...|       1.0|           [I@7f1970e1|      [1]|[Ljava.lang.Objec...|   [B@284838a9|         1|         {a=1}|            X|            X|  # noqa
```

Comment:
Hmmmmm .. I see the type is not clear here. Let me think about this a bit more.

`[B@284838a9` is a quite buggy behaviour - we should fix. So I was thinking of documenting internally since we already spent much time to figure out how it works for each case individually (at https://github.com/apache/spark/pull/20163).